### PR TITLE
Typescript

### PR DESCRIPTION
**Pull request by dhbradshaw**
This change is [Reviewable](https://reviewable.io/reviews/reviewable/nodefire/21)

### CODE REVIEW

**Review by pkaminski — CHANGES_REQUESTED** [review]
> Reviewed 13 of 13 files at r1.
> *[Reviewable](https://reviewable.io/reviews/reviewable/nodefire/21)* status: all files reviewed, 7 unresolved discussions (waiting on @dhbradshaw)

___
*[package.json, line 12 at r1](https://reviewable.io/reviews/reviewable/nodefire/21#-MHXY-W51-JAKjT279pv:-MHXY-W51-JAKjT279pw:b-snxgy) ([raw file](https://github.com/reviewable/nodefire/blob/f922ba0d3fa75637de11b9b557726b1c2627bded/package.json#L12)):*
> ```JSON
>   "scripts": {
>     "test": "node tests/runner.js",
>     "build": "tsc"
> ```

Should you add this to the dev dependencies?  I prefer not to depend on global tool installs.
___
*[built/nodefire.js, line 1 at r1](https://reviewable.io/reviews/reviewable/nodefire/21#-MHXYBav4dwGbRMomeq3:-MHXYBav4dwGbRMomeq4:b-hq6jd2) ([raw file](https://github.com/reviewable/nodefire/blob/f922ba0d3fa75637de11b9b557726b1c2627bded/built/nodefire.js#L1)):*
> ```JavaScript
> "use strict";
> ```

I wonder if there's a way to get `tsc` to emit some kind of "generated file" marker for Reviewable?  (Could be useful for our customers too.)
___
*[nodefire.ts, line 24 at r1](https://reviewable.io/reviews/reviewable/nodefire/21#-MHXTNJk4887UYXenr7g:-MHXTNJk4887UYXenr7h:b-wmc62s) ([raw file](https://github.com/reviewable/nodefire/blob/f922ba0d3fa75637de11b9b557726b1c2627bded/nodefire.ts#L24)):*
> ```XML
> export interface Reference extends admin.database.Reference {
>   childrenKeys?: Function;
>   database?: admin.database.Database;
> ```

I think every reference has a `database` reference, no need for `?`.
___
*[nodefire.ts, line 75 at r1](https://reviewable.io/reviews/reviewable/nodefire/21#-MHXTEJk5hZiaUuilDuX:-MHXTEJlBvLkKfDlB5te:bjw1t03) ([raw file](https://github.com/reviewable/nodefire/blob/f922ba0d3fa75637de11b9b557726b1c2627bded/nodefire.ts#L75)):*
> ```XML
>     this.$scope = scope || {};
>     /**
>      * @private
> ```

Do you need this on top of the `private` modifier above?
___
*[nodefire.ts, line 126 at r1](https://reviewable.io/reviews/reviewable/nodefire/21#-MHXTjcMAwRTlYh8Rr57:-MHXTjcMAwRTlYh8Rr58:b-6svi7t) ([raw file](https://github.com/reviewable/nodefire/blob/f922ba0d3fa75637de11b9b557726b1c2627bded/nodefire.ts#L126)):*
> ```XML
>    * @return {NodeFire|null} A NodeFire reference at the same location as this query or reference.
>    */
>   get ref(): NodeFire | null {
> ```

Not `| null`.
___
*[nodefire.ts, line 172 at r1](https://reviewable.io/reviews/reviewable/nodefire/21#-MHXTw614GM0_xmD7XV2:-MHXTw614GM0_xmD7XV3:b-tzoxh1) ([raw file](https://github.com/reviewable/nodefire/blob/f922ba0d3fa75637de11b9b557726b1c2627bded/nodefire.ts#L172)):*
> ```XML
>         if (_.isNil(value)) {
>           throw new Error(
>             'Missing or null variable "' + v + '" when expanding NodeFire path "' + template + '"');
> ```

Nit: could switch to an interpolated string here.
___
*[nodefire.ts, line 325 at r1](https://reviewable.io/reviews/reviewable/nodefire/21#-MHXUSPA-Go3WAzb3Bl_:-MHXUSPA-Go3WAzb3Bla:b-wmif2b) ([raw file](https://github.com/reviewable/nodefire/blob/f922ba0d3fa75637de11b9b557726b1c2627bded/nodefire.ts#L325)):*
> ```XML
>    *     pushed value (with the same scope as this object), or rejected with an error.
>    */
>   push(value: Value, options: any): Promise<NodeFire> | NodeFire {
> ```

`options: {timeout?: number}` (valid for anything that goes through `invoke()`)
___
*[nodefire.ts, line 359 at r1](https://reviewable.io/reviews/reviewable/nodefire/21#-MHXUj3s00pWktYEgGcX:-MHXUj3s00pWktYEgGcY:bg30ffz) ([raw file](https://github.com/reviewable/nodefire/blob/f922ba0d3fa75637de11b9b557726b1c2627bded/nodefire.ts#L359)):*
> ```XML
>    *     transaction committed or with undefined if it aborted, or rejected with an error.
>    */
>   transaction(updateFunction: (Value) => Value, options: {detectStuck?: number, prefetchValue?: boolean, timeout?: number}): Promise<void> {
> ```

FWIW I try to keep all source code wrapped to 100 columns.
___
*[nodefire.ts, line 497 at r1](https://reviewable.io/reviews/reviewable/nodefire/21#-MHXVotYCz6SdiYkEhxZ:-MHXVotYCz6SdiYkEhx_:b-5l05sq) ([raw file](https://github.com/reviewable/nodefire/blob/f922ba0d3fa75637de11b9b557726b1c2627bded/nodefire.ts#L497)):*
> ```XML
>     eventType: admin.database.EventType,
>     callback: (a: admin.database.DataSnapshot, b?: string) => any,
>     cancelCallback: object | ((a: Error) => any), context: object
> ```

`cancelCallback` should not be able to be an `object`.  The Firebase SDK allows this because they want it to be an optional argument but still let the client specify a `context`.  However, trying to do so in NodeFire would fail because the call to `wrapReject` below doesn't take this into account, so we just disallow it.
___
*[nodefire.ts, line 776 at r1](https://reviewable.io/reviews/reviewable/nodefire/21#-MHXXSvf-bfaiOCKZlsT:-MHXXSvf-bfaiOCKZlsU:br4dc8r) ([raw file](https://github.com/reviewable/nodefire/blob/f922ba0d3fa75637de11b9b557726b1c2627bded/nodefire.ts#L776)):*
> ```XML
>     promise = o;
>   } else if (o && typeof o.next === 'function' && typeof o.throw === 'function' && (Promise as any).co) {
>     promise = (Promise as any).co(o);
> ```

If you don't mind bumping the major version, we can get rid of support for `co` in all our libraries since we migrated to `async`/`await`.

**Review by dhbradshaw — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/reviewable/nodefire/21)* status: 10 of 13 files reviewed, 6 unresolved discussions (waiting on @pkaminski)

___
*[package.json, line 12 at r1](https://reviewable.io/reviews/reviewable/nodefire/21#-MHXY-W51-JAKjT279pv:-MHXt5ofD_Op96fU6jkp:b-896fix) ([raw file](https://github.com/reviewable/nodefire/blob/f922ba0d3fa75637de11b9b557726b1c2627bded/package.json#L12)):*
<details><summary><i>Previously, pkaminski (Piotr Kaminski) wrote…</i></summary><blockquote>

Should you add this to the dev dependencies?  I prefer not to depend on global tool installs.
</blockquote></details>

Done.
___
*[built/nodefire.js, line 1 at r1](https://reviewable.io/reviews/reviewable/nodefire/21#-MHXYBav4dwGbRMomeq3:-MHXu0xnBH1MzETIbxb4:bho6wwi) ([raw file](https://github.com/reviewable/nodefire/blob/f922ba0d3fa75637de11b9b557726b1c2627bded/built/nodefire.js#L1)):*
<details><summary><i>Previously, pkaminski (Piotr Kaminski) wrote…</i></summary><blockquote>

I wonder if there's a way to get `tsc` to emit some kind of "generated file" marker for Reviewable?  (Could be useful for our customers too.)
</blockquote></details>

Looking, I don't see one, but we could delete the built directory and add it to .gitignore and use it only for npm.
___
*[nodefire.ts, line 24 at r1](https://reviewable.io/reviews/reviewable/nodefire/21#-MHXTNJk4887UYXenr7g:-MHXll-07lGhSDiN_JvF:b-gsgsun) ([raw file](https://github.com/reviewable/nodefire/blob/f922ba0d3fa75637de11b9b557726b1c2627bded/nodefire.ts#L24)):*
<details><summary><i>Previously, pkaminski (Piotr Kaminski) wrote…</i></summary><blockquote>

I think every reference has a `database` reference, no need for `?`.
</blockquote></details>

This is beyond my current typescript fu.  The issue is that admin.database.Reference typings don't have database listed, so root and child and everything else that returns a native Reference thinks its missing a database.
___
*[nodefire.ts, line 75 at r1](https://reviewable.io/reviews/reviewable/nodefire/21#-MHXTEJk5hZiaUuilDuX:-MHXlRT6BuuhOYU1Be7-:bv82ilu) ([raw file](https://github.com/reviewable/nodefire/blob/f922ba0d3fa75637de11b9b557726b1c2627bded/nodefire.ts#L75)):*
<details><summary><i>Previously, pkaminski (Piotr Kaminski) wrote…</i></summary><blockquote>

Do you need this on top of the `private` modifier above?
</blockquote></details>

No, good catch
___
*[nodefire.ts, line 126 at r1](https://reviewable.io/reviews/reviewable/nodefire/21#-MHXTjcMAwRTlYh8Rr57:-MHXqNbg5C9Q5QC-J04D:b-896fix) ([raw file](https://github.com/reviewable/nodefire/blob/f922ba0d3fa75637de11b9b557726b1c2627bded/nodefire.ts#L126)):*
<details><summary><i>Previously, pkaminski (Piotr Kaminski) wrote…</i></summary><blockquote>

Not `| null`.
</blockquote></details>

Done.
___
*[nodefire.ts, line 325 at r1](https://reviewable.io/reviews/reviewable/nodefire/21#-MHXUSPA-Go3WAzb3Bl_:-MHXqwB77WgVuZBHhImD:b-896fix) ([raw file](https://github.com/reviewable/nodefire/blob/f922ba0d3fa75637de11b9b557726b1c2627bded/nodefire.ts#L325)):*
<details><summary><i>Previously, pkaminski (Piotr Kaminski) wrote…</i></summary><blockquote>

`options: {timeout?: number}` (valid for anything that goes through `invoke()`)
</blockquote></details>

Done.
___
*[nodefire.ts, line 497 at r1](https://reviewable.io/reviews/reviewable/nodefire/21#-MHXVotYCz6SdiYkEhxZ:-MHXsEKMAIbTdILQsJvv:b-896fix) ([raw file](https://github.com/reviewable/nodefire/blob/f922ba0d3fa75637de11b9b557726b1c2627bded/nodefire.ts#L497)):*
<details><summary><i>Previously, pkaminski (Piotr Kaminski) wrote…</i></summary><blockquote>

`cancelCallback` should not be able to be an `object`.  The Firebase SDK allows this because they want it to be an optional argument but still let the client specify a `context`.  However, trying to do so in NodeFire would fail because the call to `wrapReject` below doesn't take this into account, so we just disallow it.
</blockquote></details>

Done.
___
*[nodefire.ts, line 776 at r1](https://reviewable.io/reviews/reviewable/nodefire/21#-MHXXSvf-bfaiOCKZlsT:-MHXvHLEDj8-c9n4ELi5:b41rl3n) ([raw file](https://github.com/reviewable/nodefire/blob/f922ba0d3fa75637de11b9b557726b1c2627bded/nodefire.ts#L776)):*
<details><summary><i>Previously, pkaminski (Piotr Kaminski) wrote…</i></summary><blockquote>

If you don't mind bumping the major version, we can get rid of support for `co` in all our libraries since we migrated to `async`/`await`.
</blockquote></details>

Hmm, what's the practical cost of bumping the version?

**Review by dhbradshaw — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/reviewable/nodefire/21)* status: 10 of 13 files reviewed, 7 unresolved discussions (waiting on @pkaminski)

___
*[nodefire.js, line 499 at r2](https://reviewable.io/reviews/reviewable/nodefire/21#-MHXw58tD2mg3uefVv9f:-MHXw58tD2mg3uefVv9g:bclwrqk) ([raw file](https://github.com/reviewable/nodefire/blob/d8e2e2c8b8428358ff540f61590efbe38eb44de4/nodefire.js#L499)):*
> ```JavaScript
>   childrenKeys() {
>     return this.$ref.childrenKeys ?
>       this.$ref.childrenKeys(...arguments) :
> ```

Not sure I got this right here or below...

**Review by pkaminski — CHANGES_REQUESTED** [review]
> Reviewed 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/reviewable/nodefire/21)* status: all files reviewed, 2 unresolved discussions (waiting on @dhbradshaw)

___
*[built/nodefire.js, line 1 at r1](https://reviewable.io/reviews/reviewable/nodefire/21#-MHXYBav4dwGbRMomeq3:-MHYzK6TAhXfenwJ3pJN:b5emsn8) ([raw file](https://github.com/reviewable/nodefire/blob/f922ba0d3fa75637de11b9b557726b1c2627bded/built/nodefire.js#L1)):*
<details><summary><i>Previously, dhbradshaw (Douglas H. Bradshaw) wrote…</i></summary><blockquote>

Looking, I don't see one, but we could delete the built directory and add it to .gitignore and use it only for npm.
</blockquote></details>

OK, either way is fine.  I'm not a huge fan of the mess it makes, but don't want to make publishing unnecessarily hard either.
___
*[nodefire.js, line 499 at r2](https://reviewable.io/reviews/reviewable/nodefire/21#-MHXw58tD2mg3uefVv9f:-MHYym5PAuFlZQ62rAVC:bdbhmyt) ([raw file](https://github.com/reviewable/nodefire/blob/d8e2e2c8b8428358ff540f61590efbe38eb44de4/nodefire.js#L499)):*
<details><summary><i>Previously, dhbradshaw (Douglas H. Bradshaw) wrote…</i></summary><blockquote>

Not sure I got this right here or below...
</blockquote></details>

What do you mean?  It looks fine to me.  The weird ternary is because if we're running in encrypted mode, `firecrypt` will inject its own version of `childrenKeys` that will decrypt the keys automatically, and otherwise we need to implement it ourselves.
___
*[nodefire.ts, line 24 at r1](https://reviewable.io/reviews/reviewable/nodefire/21#-MHXTNJk4887UYXenr7g:-MHYyTor07DkcxmjvL63:b5riuxj) ([raw file](https://github.com/reviewable/nodefire/blob/f922ba0d3fa75637de11b9b557726b1c2627bded/nodefire.ts#L24)):*
<details><summary><i>Previously, dhbradshaw (Douglas H. Bradshaw) wrote…</i></summary><blockquote>

This is beyond my current typescript fu.  The issue is that admin.database.Reference typings don't have database listed, so root and child and everything else that returns a native Reference thinks its missing a database.
</blockquote></details>

I would add a comment for posterity here to explain then.
___
*[nodefire.ts, line 776 at r1](https://reviewable.io/reviews/reviewable/nodefire/21#-MHXXSvf-bfaiOCKZlsT:-MHYz27U1yH4IvALJTLp:bxonwc2) ([raw file](https://github.com/reviewable/nodefire/blob/f922ba0d3fa75637de11b9b557726b1c2627bded/nodefire.ts#L776)):*
<details><summary><i>Previously, dhbradshaw (Douglas H. Bradshaw) wrote…</i></summary><blockquote>

Hmm, what's the practical cost of bumping the version?
</blockquote></details>

Not much, just need to upgrade the dependency in `reviewable-server`.  (In principle, it would be higher it there were other users but I've never heard of any, so...)

**Review by dhbradshaw — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/reviewable/nodefire/21)* status: 1 of 9 files reviewed, 2 unresolved discussions (waiting on @pkaminski)

___
*[built/nodefire.js, line 1 at r1](https://reviewable.io/reviews/reviewable/nodefire/21#-MHXYBav4dwGbRMomeq3:-MHrqefWEQRnRVl8_wdP:b-9ek5tg) ([raw file](https://github.com/reviewable/nodefire/blob/f922ba0d3fa75637de11b9b557726b1c2627bded/built/nodefire.js#L1)):*
<details><summary><i>Previously, pkaminski (Piotr Kaminski) wrote…</i></summary><blockquote>

OK, either way is fine.  I'm not a huge fan of the mess it makes, but don't want to make publishing unnecessarily hard either.
</blockquote></details>

Ok, I set it up so that we don't keep the generated files but npm should make use of them.
___
*[nodefire.js, line 499 at r2](https://reviewable.io/reviews/reviewable/nodefire/21#-MHXw58tD2mg3uefVv9f:-MHq9BIq1jUxguehlw4r:b-8t6ik2) ([raw file](https://github.com/reviewable/nodefire/blob/d8e2e2c8b8428358ff540f61590efbe38eb44de4/nodefire.js#L499)):*
<details><summary><i>Previously, pkaminski (Piotr Kaminski) wrote…</i></summary><blockquote>

What do you mean?  It looks fine to me.  The weird ternary is because if we're running in encrypted mode, `firecrypt` will inject its own version of `childrenKeys` that will decrypt the keys automatically, and otherwise we need to implement it ourselves.
</blockquote></details>

I was unsure whether `...arguments` could be replaced by `options`.  Based on experimentation, it looks like it can.
___
*[nodefire.ts, line 24 at r1](https://reviewable.io/reviews/reviewable/nodefire/21#-MHXTNJk4887UYXenr7g:-MHpz2aNC1-rL6-ztktO:b-896fix) ([raw file](https://github.com/reviewable/nodefire/blob/f922ba0d3fa75637de11b9b557726b1c2627bded/nodefire.ts#L24)):*
<details><summary><i>Previously, pkaminski (Piotr Kaminski) wrote…</i></summary><blockquote>

I would add a comment for posterity here to explain then.
</blockquote></details>

Done.
___
*[nodefire.ts, line 776 at r1](https://reviewable.io/reviews/reviewable/nodefire/21#-MHXXSvf-bfaiOCKZlsT:-MHqAOpjAeZptLQ-JJlY:bzejql) ([raw file](https://github.com/reviewable/nodefire/blob/f922ba0d3fa75637de11b9b557726b1c2627bded/nodefire.ts#L776)):*
<details><summary><i>Previously, pkaminski (Piotr Kaminski) wrote…</i></summary><blockquote>

Not much, just need to upgrade the dependency in `reviewable-server`.  (In principle, it would be higher it there were other users but I've never heard of any, so...)
</blockquote></details>

Maybe I'll punt on this.  I should probably wrap this typescript work up.

**Review by pkaminski — CHANGES_REQUESTED** [review]
> Reviewed 9 of 16 files at r3.
> *[Reviewable](https://reviewable.io/reviews/reviewable/nodefire/21)* status: all files reviewed, 3 unresolved discussions (waiting on @dhbradshaw)

___
*[package-lock.json, line 0 at r3](https://reviewable.io/reviews/reviewable/nodefire/21#-MHs5njI1b2b67rmLFiU:-MHs5njI1b2b67rmLFiV:b-noj5u4) ([raw file](https://github.com/reviewable/nodefire/blob/3448d7579e94178f1172341e9622d6b3cb7fded9/package-lock.json#L0)):*
We use `yarn` for package management and locking, not `npm`.  If creating this file is unavoidable please add to `.gitignore`.
___
*[src/index.ts, line 2 at r3](https://reviewable.io/reviews/reviewable/nodefire/21#-MHs67gcCcifAUyJeCHn:-MHs67gcCcifAUyJeCHo:b6egrcj) ([raw file](https://github.com/reviewable/nodefire/blob/3448d7579e94178f1172341e9622d6b3cb7fded9/src/index.ts#L2)):*
> ```XML
> import Nodefire from './nodefire';
> export default Nodefire;
> ```

FYI, IIUC you can just `export * from './nodefire'`.
___
*[src/nodefire.ts, line 327 at r3](https://reviewable.io/reviews/reviewable/nodefire/21#-MHs6wtp7b5Df2OLeDPV:-MHs6wtq-6CqgzOsIccm:bl50fxy) ([raw file](https://github.com/reviewable/nodefire/blob/3448d7579e94178f1172341e9622d6b3cb7fded9/src/nodefire.ts#L327)):*
> ```XML
>   push(value: Value, options?: { timeout?: number }): Promise<NodeFire> {
>     if (value === undefined || value === null) {
>       return Promise.resolve(new NodeFire(this.$ref.push(), this.$scope));
> ```

Wow, that bug has been there for a long time!  A win for type checking, I suppose.  :)
___
*[src/nodefire.ts, line 710 at r3](https://reviewable.io/reviews/reviewable/nodefire/21#-MHs7G3h-xEpvdfs_LLx:-MHs7G3h-xEpvdfs_LLy:b-3d83si) ([raw file](https://github.com/reviewable/nodefire/blob/3448d7579e94178f1172341e9622d6b3cb7fded9/src/nodefire.ts#L710)):*
> ```XML
> wrapNodeFire('startAt');
> wrapNodeFire('endAt');
> wrapNodeFire('equalTo');
> ```

You should also remove this line since you re-implemented above.

**Review by dhbradshaw — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/reviewable/nodefire/21)* status: 2 of 6 files reviewed, 2 unresolved discussions (waiting on @pkaminski)

___
*[package-lock.json, line  at r3](https://reviewable.io/reviews/reviewable/nodefire/21#-MHs5njI1b2b67rmLFiU:-MHv6XRS-cOuVH6t4gBk:bmzp7zu) ([raw file](https://github.com/reviewable/nodefire/blob/3448d7579e94178f1172341e9622d6b3cb7fded9/package-lock.json#L)):*
<details><summary><i>Previously, pkaminski (Piotr Kaminski) wrote…</i></summary><blockquote>

We use `yarn` for package management and locking, not `npm`.  If creating this file is unavoidable please add to `.gitignore`.
</blockquote></details>

Sorry, not unavoidable.  It was an error to include that.
___
*[src/index.ts, line 2 at r3](https://reviewable.io/reviews/reviewable/nodefire/21#-MHs67gcCcifAUyJeCHn:-MHutWdq-y0grTq9E9Dz:bb7kz30) ([raw file](https://github.com/reviewable/nodefire/blob/3448d7579e94178f1172341e9622d6b3cb7fded9/src/index.ts#L2)):*
<details><summary><i>Previously, pkaminski (Piotr Kaminski) wrote…</i></summary><blockquote>

FYI, IIUC you can just `export * from './nodefire'`.
</blockquote></details>

I've added that line in so that the consumer can access non-default exports.  But it doesn't seem to work on its own.  I'm not sure why.
___
*[src/nodefire.ts, line 710 at r3](https://reviewable.io/reviews/reviewable/nodefire/21#-MHs7G3h-xEpvdfs_LLx:-MHupAOgAhmEOC-rZ-Os:b-896fix) ([raw file](https://github.com/reviewable/nodefire/blob/3448d7579e94178f1172341e9622d6b3cb7fded9/src/nodefire.ts#L710)):*
<details><summary><i>Previously, pkaminski (Piotr Kaminski) wrote…</i></summary><blockquote>

You should also remove this line since you re-implemented above.
</blockquote></details>

Done.

**Review by pkaminski — APPROVED** [review]
> Reviewed 6 of 6 files at r4.
> *[Reviewable](https://reviewable.io/reviews/reviewable/nodefire/21)* status: :shipit: complete! all files reviewed, all discussions resolved

**Review by pkaminski — APPROVED** [review]
> Reviewed 2 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/reviewable/nodefire/21)* status: :shipit: complete! all files reviewed, all discussions resolved
